### PR TITLE
Fixes #65: Add Schema.Lazy to support recursive macro derivation

### LINKED ISSUE #65: Support for recursive types
The current encoding of case classes has a field for each field of the case class, that returns a tuple of `String` and `Schema`. But if the type is recursive or if indirectly it contains classes that have circular dependencies, it will blow up with stack overflow. Example: https://scastie.scala-lang.org/mSCrN0pQRjOw9pBZNzMSAQ

In Caliban I dealt with this by making the field information lazy (`() => FieldInfo` instead of `FieldInfo`).

### PR DESCRIPTION
**Pull request by thinkharderdev**
Fixes #65 Handle recursive macro derivation by wrapping case class fields in a `Schema.Lazy[A](f: () => Schema[A])`

### CODE REVIEW

File: zio-schema/shared/src/main/scala/zio/schema/Schema.scala [hunk 1]
```diff
@@ -858,6 +860,8 @@ object Schema {
 
   final case class EitherSchema[A, B](left: Schema[A], right: Schema[B]) extends Schema[Either[A, B]]
 
+  final case class Lazy[A](schema: () => Schema[A]) extends Schema[A]
```
Comment (jdegoes) [inline zio-schema/shared/src/main/scala/zio/schema/Schema.scala:863]:
```suggestion
  final case class Lazy[A](private val schema0: () => Schema[A]) extends Schema[A] {
    lazy val schema = schema0
  }
```

File: zio-schema/shared/src/test/scala/zio/schema/DeriveSchemaSpec.scala [hunk 2]
```diff
@@ -28,61 +30,82 @@ object DeriveSchemaSpec extends DefaultRunnableSpec {
...
+              (s: Status) => s.asInstanceOf[Pending.type]
+            )
           )
-        )
 
-      assert(derived)(hasSameSchema(expected))
-    }
+        assert(derived)(hasSameSchema(expected))
+      },
+      test("correctly derives for case classes with recursive types") {
```
Comment (jdegoes) [inline zio-schema/shared/src/test/scala/zio/schema/DeriveSchemaSpec.scala:100]: So great to see these passing! 🎉 👯‍♂️

**Review by jdegoes — CHANGES_REQUESTED** [review]
Just a minor comment, otherwise looks great!

File: zio-schema/shared/src/main/scala/zio/schema/Schema.scala (hunk 1 above)
Comment (thinkharderdev) [reply zio-schema/shared/src/main/scala/zio/schema/Schema.scala:863]: Good call!

Comment (thinkharderdev) [reply zio-schema/shared/src/main/scala/zio/schema/Schema.scala]: Done!

**Review by jdegoes — APPROVED** [review]

**Comment by jdegoes** [comment]
Fantastic work on this. Thank you! 🙏

**Comment by jdegoes** [comment]
@thinkharderdev Looks like docs failed, can you take a look?

<https://github.com/zio/zio-schema/actions/runs/909896564>

**Comment by thinkharderdev** [comment]
> @thinkharderdev Looks like docs failed, can you take a look?
> 
> https://github.com/zio/zio-schema/actions/runs/909896564

Should be fixed with my PR. Never added a website to this repo.

**Pull request merged**